### PR TITLE
Speed up pow on CPU for non-contiguous float tensors by using the correct type

### PR DESCRIPTION
**Pull request by xuhdev**
In pow_tensor_scalar_kernel, currently `exp` is always double. When base is float, this makes
the call to `std::pow(base, exp)` (post-change line 88) uses an overload that is unoptimized because `exp` is double.

Some benchmarks:

```python
import timeit

for n, t in [(1000, 20000),
             (300_000, 400)]:
    for dtype in ('torch.float', 'torch.double'):
        print(f'a.pow(b) (a.numel() == {n}) for {t} times')
        print(f'dtype {dtype}, {t} times')
        print('Contiguous', end='\t')
        print(timeit.timeit(f'a.pow(1.1); a.pow(0.2)',
                            setup=f'import torch; a = torch.arange({n},
device="cpu", dtype={dtype})',
                            number=t))

        # Non-contiguous
        print('Non-contiguous', end='\t')
        print(timeit.timeit(f'a.pow(1.1); a.pow(0.2)',
                            setup=f'import torch; a = torch.full((100,
{n//100}), 1.1, device="cpu", dtype={dtype}).t()',
                            number=t))
```

Results:

before:

```
a.pow(b) (a.numel() == 1000) for 20000 times
dtype torch.float, 20000 times
Contiguous  1.9101960129992221
Non-contiguous  5.76372037299916
a.pow(b) (a.numel() == 1000) for 20000 times
dtype torch.double, 20000 times
Contiguous  2.725056825000138
Non-contiguous  5.905541740999979
a.pow(b) (a.numel() == 300000) for 400 times
dtype torch.float, 400 times
Contiguous  4.626354793999781
Non-contiguous  27.473886219999258
a.pow(b) (a.numel() == 300000) for 400 times
dtype torch.double, 400 times
Contiguous  9.953559645999121
Non-contiguous  28.30575869799941
```

after:

```
a.pow(b) (a.numel() == 1000) for 20000 times
dtype torch.float, 20000 times
Contiguous  1.8944922199989378
Non-contiguous  4.036399689000973
a.pow(b) (a.numel() == 1000) for 20000 times
dtype torch.double, 20000 times
Contiguous  2.708891585998572
Non-contiguous  5.898831417998736
a.pow(b) (a.numel() == 300000) for 400 times
dtype torch.float, 400 times
Contiguous  4.557184483001038
Non-contiguous  17.10840253800052
a.pow(b) (a.numel() == 300000) for 400 times
dtype torch.double, 400 times
Contiguous  9.803347787001258
Non-contiguous  28.14479073000075
```

Note the significant improvement (~30%) for non-contiguous float tensors.

(Additionally, due to the movement of assignment to `exp`, this commit
also had the chance to simplify a bit of the code that handles of the
situation in which base is integral.)

### CODE REVIEW

**Comment by VitalyFedyunin** [comment]
@pbelevich can you please take a look

**Comment by pbelevich** [comment]
I have no objections, if we manage to fix pow_test.cpp

**Review by pbelevich — CHANGES_REQUESTED** [review]
I have no objections but pow_test.cpp should be fixed

**Comment by xuhdev** [comment]
@pbelevich Sorry for the delay -- I just pushed the fix and now we are waiting for the CI. The reason it failed before was because `Scalar` does not support `to<long double>`.

@pytorchbot retest this please

**Comment by xuhdev** [comment]
It still failed, looking into it now..

**Comment by xuhdev** [comment]
I found a potential issue: Scalar never tells us whether the passed in Scalar is a double or float. Is there a way to get this info? Otherwise, this patch will make a call of `pow(float, double)` to be effectively `pow(float, float)`, a potential accuracy loss that is not intended by the user.

**Comment by pbelevich** [comment]
@xuhdev Scalar::isIntegral/isFloatingPoint/...

**Comment by xuhdev** [comment]
@pbelevich The issue is not about floating-point or integer, but to distinguish `double` and `float`...

**Comment by VitalyFedyunin** [comment]
Scalar is always `double` or `int64_t` or complex ( see `c10/core/Scalar.h` )

**Comment by VitalyFedyunin** [comment]
@xuhdev taking into account that scalar is double, do we still want  this PR?

**Comment by xuhdev** [comment]
@VitalyFedyunin Probably no. Let me close this. Thanks!
